refactor(appservice): Better virtual client sync

agraven opened the pull request:
Store the membership state of clients in the appservice's namespace, and
construct sync events based on that information

poljar commented on crates/matrix-sdk-appservice/src/lib.rs:
```diff
@@ -627,6 +632,89 @@ impl AppService {
+                continue;
+            }
+            let localpart = event.state_key().localpart();
+            client
```
We'll probably want to store these as a single transaction instead of saving each member separately. A TODO + open issue will do.

agraven replied:
Is that possible when I don't have access to the private methods in the `matrix-sdk` crate?

poljar replied:
It's not possible currently, there's now an issue where we can add what else we would like from the store: https://github.com/matrix-org/matrix-rust-sdk/issues/730.

poljar commented on crates/matrix-sdk-appservice/src/lib.rs:
```diff
@@ -627,6 +632,89 @@ impl AppService {
+        }
+
+        // Send a sync event to each active client
+        // TODO: What do we do with state events?
```
Could you expand on this? What is the exact problem with state events?

agraven replied:
I was uncertain if putting any state events received in the transaction into the `state` section of the synthesized sync event would be needed for the virtual clients to maintain a coherent view of the rooms they're participating in. Though the `state` section is for events from *before* the bits in `timeline`, so that adds an additional layer of complexity. I'm a bit unclear on what the right move is here.

poljar replied:
Since the major difference between what's in `state` vs what's in the `timeline` is the former will not be displayed to the user while the later will, and the app-service doesn't display any of this to users, we can safely just put everything into the timeline.

We just need to make sure that, what we would otherwise put into the `state` section, comes before anything that belongs into the `timeline` section.

I think this is already the case?

agraven replied:
I believe so, yes. In that case I don't think there's any need to change anything here. Thanks for helping me think through this.

poljar replied:
Let's just remove the TODO if there's nothing to do here. :)

poljar commented on crates/matrix-sdk-appservice/src/lib.rs:
```diff
@@ -627,6 +632,89 @@ impl AppService {
+            };
+            let mut response = sync_events::v3::Response::new(transaction.txn_id.to_string());
+
+            for raw_event in &transaction.events {
```
So the events app services receive are received in a flattened out structure. Clients on the other hand receive them in individual room buckets.

The important thing here is that each individual room  is in a bucket that denotes the room membership state of the current user. This bucketing is something the client expects.

Could we add a comment explaining this?

Also, this will modify and create personalized sync responses for each individual virtual user, has this been tried out in the real world, will this be slow? How many virtual users does an app-service usually handle?

It also seems that if we decide to do this we could, or rather should, parallelize this so each virtual client is handled in a task. Again, a TODO + issue would be enough.

agraven replied:
The main reasoning for taking the one sync response per virtual client approach was that it was the only way I could think of handler per-user membership properly without making deeper changes in `matrix-sdk`. It might be possible to reduce overhead by making each virtual user use the same store or something, but I'm assuming an approach in that direction would require deeper changes within the matrix-sdk crate to make that work properly

poljar replied:
Yeah, I assumed as much and that's understandable. Though would be nice to know if it has been tried out and if so under what load vs what an usual load for app-services is.

agraven replied:
Regarding parallelism, I'm thinking the `futures` crate's [`StreamExt::for_each_concurrent`](https://docs.rs/futures/latest/futures/stream/trait.StreamExt.html#method.for_each_concurrent) would be the easiest way of introducing concurrency, but I don't know if tokio is able to introduce maximum parallelism when using that, or if tasks are needed for that.

agraven replied:
Hm, making the parallelism work properly with lifetimes is turning out more complicated than expected, I'll leave it as a TODO so the focus can be on the basic behavior

poljar replied:
Sounds good.

johannescpk replied:
> Though would be nice to know if it has been tried out and if so under what load vs what an usual load for app-services is.

I'm not aware of any benchmarking or tests with large amounts of virtual users - so it's hard to say. However, I'd say that the solution outlined here already helps for the basic usage pattern and shows some of the needs for a solution that gets more deeply embedded into the sdk. Going by that I'd advocate for getting this merged so we can collect some real world experiences and then refactored before a crate release.

poljar commented on crates/matrix-sdk/src/client/mod.rs:
```diff
@@ -266,6 +266,22 @@ impl Client {
+    ///
+    /// [transaction]: https://matrix.org/docs/spec/application_service/r0.1.2#put-matrix-app-v1-transactions-txnid
+    #[cfg(feature = "appservice")]
+    pub async fn receive_preprocessed_transaction(
```
Does this obsolete the `receive_transaction()` method? If not could you add some info how this method differs from the `receive_transaction()` one, cross linking from one to the other would be great as well.

agraven replied:
I hadn't considered that, there probably isn't much point in keeping the old version of the method

poljar replied:
Let's remove it then.

poljar reviewed (changes requested):
Seems sensible from what I understand, left a couple of nits and questions. @johannescpk mind taking a look at this one as well?

jplatte commented on crates/matrix-sdk-appservice/src/lib.rs:
```diff
@@ -627,6 +632,89 @@ impl AppService {
+                    Some(value) => String::from_utf8(value).ok().map(MembershipState::from),
+                    None => None,
+                };
+                let json = Raw::into_json(raw_event.clone());
+
+                match membership.unwrap_or(MembershipState::Join) {
+                    MembershipState::Join => {
+                        let room = response.rooms.join.entry(room_id).or_default();
+                        room.timeline.events.push(Raw::from_json(json))
+                    }
+                    MembershipState::Leave | MembershipState::Ban => {
+                        let room = response.rooms.leave.entry(room_id).or_default();
+                        room.timeline.events.push(Raw::from_json(json))
```
```suggestion

                match membership.unwrap_or(MembershipState::Join) {
                    MembershipState::Join => {
                        let room = response.rooms.join.entry(room_id).or_default();
                        room.timeline.events.push(raw_event.clone().cast());
                    }
                    MembershipState::Leave | MembershipState::Ban => {
                        let room = response.rooms.leave.entry(room_id).or_default();
                        room.timeline.events.push(raw_event.clone().cast());
```

agraven replied:
Thanks, I felt like there had to be a better way of doing that.

johannescpk commented:
Can we add a comment to the crate's top-level documentation under "Status" mentioning that

> The crate relies on the registration being always in sync with the actual registration used by synapse. That's because it's required for the access tokens and because membership states for virtual users are determined based on the registered namespace.
> 
> Warning: Non-exclusive registration namespaces are not yet supported and hence might lead to undefined behavior.

Also filed an upstream issue that we can deprecate/remove the helper method we had in ruma: https://github.com/ruma/ruma/issues/1159.

Other than that this LGTM.

agraven commented:
The parallelism turned out to be easier than i thought after giving it a second go. I believe I've addressed all review feedback at this point, and that this should be ready for a final review pass

johannescpk commented on crates/matrix-sdk-appservice/src/lib.rs:
```diff
@@ -29,6 +29,16 @@
 //! The crate is in an experimental state. Follow
 //! [matrix-org/matrix-rust-sdk#228] for progress.
 //!
+//! The crate relies on the registration being always in sync with the actual
```
I'd put it into its own "Registration" section, as this likely will keep being an requirement regardless of the crate status.

johannescpk commented on crates/matrix-sdk-appservice/src/lib.rs:
```diff
@@ -29,6 +29,16 @@
 //! [matrix-org/matrix-rust-sdk#228] for progress.
 //!
+//! The crate relies on the registration being always in sync with the actual
+//! registration used by synapse. That's because it's required for the access
```
```suggestion
//! registration used by homeserver. That's because it's required for the access
```

johannescpk commented on crates/matrix-sdk-appservice/src/lib.rs:
```diff
@@ -29,6 +29,16 @@
+//! tokens and because membership states for virtual users are determined based
+//! on the registered namespace.
+//!
+//! ## Warning
```
I think we can remove the "Warning" heading here, as it's just an additional remark.

agraven replied:
Done

poljar reviewed (approved):
Looks good, thanks.
